File: server/middleware/rateLimit.js
// Minimal in-memory per-IP rate limiter for the unauthenticated public
// widget endpoint. Good enough for a POC; not meant to survive a restart
// or run across multiple server instances.
function rateLimit({ windowMs, max }) {
  const hits = new Map(); // ip -> array of request timestamps

  return function (req, res, next) {
    const ip = req.ip || req.socket.remoteAddress || "unknown";
    const now = Date.now();
    const recent = (hits.get(ip) || []).filter((ts) => now - ts < windowMs);

    if (recent.length >= max) {
      return res.status(429).json({
        error: { title: "TooManyRequests", message: "Please wait a moment before submitting again." },
      });
    }

    recent.push(now);
    hits.set(ip, recent);
    next();
  };
}

module.exports = { rateLimit };

File: fastapi-app/app/rate_limit.py
import time
from collections import defaultdict

from fastapi import Request, HTTPException, status

# Minimal in-memory per-IP rate limiter for the unauthenticated public widget
# endpoint. Good enough for a POC; does not survive a restart and does not
# work across multiple worker processes.
_hits: dict[str, list[float]] = defaultdict(list)


def rate_limit(window_seconds: int = 60, max_requests: int = 5):
    def _check(request: Request):
        ip = request.client.host if request.client else "unknown"
        now = time.time()
        recent = [ts for ts in _hits[ip] if now - ts < window_seconds]

        if len(recent) >= max_requests:
            raise HTTPException(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                detail={
                    "title": "TooManyRequests",
                    "message": "Please wait a moment before submitting again.",
                },
            )

        recent.append(now)
        _hits[ip] = recent

    return _check
